### PR TITLE
Allow `whodunnit` to be a proc

### PR DESCRIPTION
**Pull request by narfanator**
Basically, I've got a circumstance where I'm trying to figure out *what* is changing one of my models (it's not Sidekiq jobs, and it's not going through the controllers...).

Right now I've got:

```
class PapertrailSidekiqMiddleware
  def call(worker_class, job, queue)
    if(worker_class.class == Sidekiq::Extensions::DelayedClass)
      PaperTrail.whodunnit = YAML.load(job["args"].first).first.to_s
    else
      PaperTrail.whodunnit = worker_class.class.to_s
    end
    yield
  end
end

Sidekiq.configure_server do |config|
  config.server_middleware do |chain|
    chain.add PapertrailSidekiqMiddleware
  end
end
```

This works, but it also only covers Sidekiq; if there's no place to stick the `whodunnit` initialization like this, there's no way to make the `whodunnit` dynamic and record the actual situation.

However, with a change like this, you could set a whodunnit during the creation of the paper trail, so you could evaluate anything you like (inspecting the callstack, the originally run Ruby command, etc etc).

PS - Honestly, I was 100% convinced I wanted this feature as part of tracking down this issue, but in this moment I'm honestly no remembering what the plan was that needed it.... :P

PPS - I totally don't expect this PR to get accepted, but I wanted to start the conversation and figured this'd be more illustrative than making an issue. LMK how I should proceed!

PPPS -  I also ran into a couple of issues when first running `rspec` on Papertrail, so there's a couple of Gemfile commits to fix that.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,2 +1,5 @@
 source "https://rubygems.org"
 gemspec
+
+gem "rubocop-rspec", "~> 1.15.1"
+gem "rails-controller-testing"
```
Comment (jaredbeck) [inline Gemfile]: Please do not edit the `Gemfile`. Please edit `Appraisals` instead. But, I don't think any changes to gems are necessary.

File: lib/paper_trail.rb [hunk 2]
```diff
@@ -112,7 +112,11 @@ def whodunnit(value = nil)
           paper_trail_store[:whodunnit] = previous_whodunnit
         end
       else
-        paper_trail_store[:whodunnit]
+        if paper_trail_store[:whodunnit].class == Proc
+          paper_trail_store[:whodunnit].call
+        else
+          paper_trail_store[:whodunnit]
+        end
```
Comment (jaredbeck) [inline lib/paper_trail.rb]: Should the conditional check `respond_to?(:call)` instead of `== Proc`?

File: spec/paper_trail_spec.rb [hunk 3]
```diff
@@ -101,6 +101,15 @@
...
     end
+
+    context "when set to a proc" do
+      it "evaluates the proc each time a version is made" do
+        call_count = 0
+        described_class.whodunnit = proc { call_count += 1 }
+        expect(described_class.whodunnit).to eq(1)
+        expect(described_class.whodunnit).to eq(2)
+      end
+    end
```
Comment (jaredbeck) [inline spec/paper_trail_spec.rb]: Nice test :+1:

Comment (narfanator) [inline spec/paper_trail_spec.rb]: :D

File: lib/paper_trail.rb (hunk 2 above)
Comment (narfanator) [inline lib/paper_trail.rb]: Yah, that makes a lot of sense; sharp knives and all that.

File: Gemfile (hunk 1 above)
Comment (narfanator) [inline Gemfile]: Kk. I'm not familiar with `Appraisals`, but I can go figure it out.

Comment (narfanator) [inline Gemfile]: You're correct, when run via `appraisal`, everything is fine.

**Comment by narfanator** [comment]
Updated as per review and failed tests; rewrote Git history of this branch to keep it super clean.

I added a bit to the documentation where I (as a new and inexperienced contributor) was confused.

**Comment by jaredbeck** [comment]
Merged. Thanks for your contribution!

**Pull request merged**